doc/protocol.md

```diff
@@ -829,6 +829,119 @@ or pipe is done.
+   entries to be reported.  Use `*` to report all files.  Can also be used to
+   implement search (`searc*`) or hide temporary files (`[!.]*`).
+ * `close-on-error` (optional, default: true): if the channel should close
+   (with `problem`) on failure to `open()` the named `path`
+ * `watch` (optional, default: false): if the channel should watch for changes
+ * `follow` (default: true): if the trailing component of `path` is a symbolic
+   link, whether we should follow it.  `no-follow` mode is not supported for
+   `watch` (but is also mostly unnecessary).
+
+A channel is created with an associated absolute pathname ('path').  A path is
+a string that you pass to the `open()` syscall.  The object returned by that
+call is described by the (JSON Document) value of the channel.
+
+Strictly speaking, the path is specified to the channel as a unicode string,
+but paths on the filesystem are binary strings.  This mapping is handled using
+the usual Python convention: filenames are stored in utf8 and encoded and
+decoded with `errors='surrogateescape'`, which allows for handling of filenames
+which are not utf8.
+
+Conceptually, the channel represents the value of a JSON object describing what
+is found at the path, or the reason for an error.  Each message on the channel
+is a JSON Merge Patch (RFC7396) which updates or replaces this object.
+
+The top-level keys in the object are:
+
+ - `info`: the info about the reported path (described below)
+ - `problem`: a "problem code" like `not-found` or `not-directory`
+ - `message`: a user-readable (`strerror`) string describing the error
+ - `partial`: `true` if the data is in a "partial" state.  This is used when
+   sending updates in multiple pieces.
+
+If `close-on-error` is set, `problem` and `message` will never be sent, but the
+channel will be closed with the `problem` code set to indicate the error.
+
+Otherwise, the value of the `info` attribute is a JSON object describing the
+attributes of the file.  Any attribute which cannot be read (eg: due to
+permissions problems) is simply not reported.
+
+ * `type`: a string: `file`, `directory`, `link`, or `special`
+ * `mode`: an integer representing the "file permission" bits.  As per JSON,
+   this is transmitted as a decimal value, but it is meant to be interpreted in
+   octal, in the usual way.
+ * `uid`: an integer, the uid of the file owner (`st_uid`)
+ * `owner`: a string, or an integer if the lookup failed
+ * `gid`: an integer, the gid of the file group (`st_gid`)
+ * `group`: a string, or an integer if the lookup failed
+ * `size`: an integer, the (apparent) size of the file (`st_size`)
+ * `modified`: a float, the mtime of the file (`st_mtim`)
+ * `tag`: a value type: the current 'transaction tag' of the file, with the
+   same meaning as elsewhere in this document.  Ideally: this changes if the
+   content of the file changes.  This is stable between different channels.
```

Comment:
What does "stable between different channels" mean? fswatch, fsinfo, fsread will report the same tag for the same file state?

Reply:
Yes.  Do you have a suggestion for better wording?

Reply:
hmm.. "This is the same as in fsread1 and fsreplace1"? But no strong opinion, feel free to resolve this. I mostly just wanted to check that I understood it right.